db/test/helper.js

```diff
@@ -223,3 +231,216 @@ exports.assertNoTableColumn = async (table, column) => {
+      await exports.upgradeTo(PREV_VERSION);
+      await exports.withDbClient(createData);
+    });
+
+    teardown(async function() {
+      runOnlineBatches.resetHooks();
+    });
+
+    test('successful upgrade, downgrade process', async function() {
+      await exports.withDbClient(async client => {
+        await startCheck(client);
+        await withCheckpoints('up', {
+          start: async () => await concurrentCheck(client),
+          preOnline: async () => await concurrentCheck(client),
+          midOnline: async () => await concurrentCheck(client),
+          done: async () => await concurrentCheck(client),
+        });
+        await finishedCheck(client);
+        await withCheckpoints('down', {
+          start: async () => await concurrentCheck(client),
+          preOnline: async () => await concurrentCheck(client),
+          midOnline: async () => await concurrentCheck(client),
+          done: async () => await concurrentCheck(client),
+        });
+        await startCheck(client);
+
+        // these are checks to make sure that there are sufficient batches of online
+        // migration and downgrade for the tests to be effective; if these fail, add more
+        // test data.  Set DEBUG=showProgress to help debugging.
+        if (onlineMigration) {
+          assert(sawMigrationBatches, 'did not see multiple batches of online migration');
+        }
+        if (onlineDowngrade) {
+          assert(sawDowngradeBatches, 'did not see multiple batches of online downgrade');
+        }
+      });
+    });
+
+    // remainder of the tests are only interesting with an online migration and downgrade
+    if (!onlineMigration && !onlineDowngrade) {
+      return;
+    }
+
+    test('upgrade fails mid-online, restarted, downgrade fails mid-online, restarted', async function() {
+      await exports.withDbClient(async client => {
+        await startCheck(client);
+        await withCheckpoints('up', {
+          midOnline: async () => { throw abort; },
+        });
+        await withCheckpoints('up', {
+          start: async () => await concurrentCheck(client),
+          preOnline: async () => await concurrentCheck(client),
+          midOnline: async () => await concurrentCheck(client),
+          done: async () => await concurrentCheck(client),
+        });
+        await finishedCheck(client);
+        await withCheckpoints('down', {
+          midOnline: async () => { throw abort; },
+        });
+        await withCheckpoints('down', {
+          start: async () => await concurrentCheck(client),
+          preOnline: async () => await concurrentCheck(client),
+          midOnline: async () => await concurrentCheck(client),
+          done: async () => await concurrentCheck(client),
+        });
+        await startCheck(client);
+        assert(sawMigrationBatches, 'did not see multiple batches of online migration');
+        assert(sawDowngradeBatches, 'did not see multiple batches of online downgrade');
```

Comment:
@djmitche I'm working on the online migration for the tasks table for #3578, and trying to use this helper for the migration tests, but I'm not sure I understand this check. Since I'm trying to test only the online upgrade part, I set

```js
helper.dbVersionTest({
    version: THIS_VERSION,
    onlineMigration: true,
    onlineDowngrade: false,
...
```

But all these tests that assert unconditionally on both `sawMigrationBatches` and `sawDowngradeBatches` fail, and with reason I think, since I've not implemented an online downgrade for this db version.

I think I would expect these lines to be like lines 360-365:

```js
if (onlineMigration) {
    assert(sawMigrationBatches, 'did not see multiple batches of online migration');
}
if (onlineDowngrade) {
     assert(sawDowngradeBatches, 'did not see multiple batches of online downgrade');
}
```

What am I missing here?
(Also, I don't know if adding a comment here is an appropriate venue for this)